### ISSUE TITLE
Type inference finfun

### ISSUE THREAD

**Issue by thery** [math-comp/math-comp#370]
It seems that in 1.9.0 type inference has changed for finfun.

Before this code was working :
```
From mathcomp Require Import all_ssreflect.
Check (fun n (f : {ffun 'I_n -> _}) i => f i = i). 
```
now there is an error :

```The term "Phant ('I_n -> ?T)" has type
 "phant ('I_n -> ?T)"
while it is expected to have type
 "phant
    (forall x : ordinal_finType n, ?rT x)"
(cannot instantiate "?T" because 
"x" is not in its scope: available arguments
are "n").
```

but this still works :
```
Check (fun n (f : {ffun _ -> 'I_n}) i => f i = i). 
```

Is this the expected behaviour?

**Comment by ggonthier** [comment]
This is a consequence of a Coq deficiency, that was exposed by the extension of `finfun` to dependent functions in version 1.9.0. It is documented in the `finfun.v` header, which includes some redundant definitions mitigating the issue - although it does not mention the issue you encountered, as specifying complete types is good practice generally followed in the MathComp library.
   The Coq issue coq/coq#9663  has been resolved, so the pre-1.9.0 behaviour should eventually be restored by upgrading Coq.

**Comment by thery** [comment]
ok, tested it with 8.10+beta1, it works!
